core/textile/sync/synchronizer.go

```diff
@@ -0,0 +1,304 @@
+	}()
+}
+
+// Restores a previously initialized queue
+func (s *synchronizer) RestoreQueue() error {
+	if err := s.restoreQueue(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *synchronizer) startSyncLoop(ctx context.Context, queue *list.List) {
+	queueMutex := s.queueMutexMap[queue]
+	// Initial sync
+	queueMutex.Lock()
+	s.sync(ctx, queue)
+	queueMutex.Unlock()
+
+	for {
+		queueMutex.Lock()
+		timeAfterNextSync := 30 * time.Second
+
+		select {
+		case <-time.After(timeAfterNextSync):
+			s.sync(ctx, queue)
+
+		case <-s.syncNeeded:
+			s.sync(ctx, queue)
+
+		// Break execution in case of shutdown
+		case <-ctx.Done():
+			queueMutex.Unlock()
+			return
+		case <-s.shuttingDown:
+			queueMutex.Unlock()
+			return
+		}
+
+		queueMutex.Unlock()
+	}
+}
+
+func (s *synchronizer) Shutdown() {
+	s.shuttingDown <- true
+}
+
+var errMaxRetriesSurpassed = errors.New("max retries surpassed")
+
+func (s *synchronizer) executeTask(ctx context.Context, t *Task) error {
+	var err error
+
+	switch t.Type {
+	case addItemTask:
+		err = s.processAddItem(ctx, t)
+	case removeItemTask:
+		err = s.processRemoveItem(ctx, t)
+	case pinFileTask:
+		err = s.processPinFile(ctx, t)
+	case unpinFileTask:
+		err = s.processUnpinFile(ctx, t)
+	case createBucketTask:
+		err = s.processCreateBucket(ctx, t)
+	case bucketBackupOnTask:
+		err = s.processBucketBackupOn(ctx, t)
+	case bucketBackupOffTask:
+		err = s.processBucketBackupOff(ctx, t)
+	default:
+		log.Warn("Unexpected action on Textile sync, executeTask")
+	}
+
+	if err != nil {
+		t.State = taskFailed
+		t.Retries++
+
+		// Remove from queue if it surpassed the max amount of retries
+		if t.MaxRetries != -1 && t.Retries > t.MaxRetries {
+			t.State = taskDequeued
+			return errMaxRetriesSurpassed
+		}
+
+		// Retry task
+		t.State = taskQueued
+	} else {
+		t.State = taskSucceeded
+	}
+
+	return err
+}
+
+var parallelTaskCount = 0
+
+func (s *synchronizer) sync(ctx context.Context, queue *list.List) error {
+	queueName := "buckets"
+	if queue == s.filePinningQueue {
+		queueName = "file pinning"
+	}
+
+	log.Debug(fmt.Sprintf("Textile sync [%s]: Sync start", queueName))
+	s.printQueueStats(queue)
+
+	for curr := queue.Front(); curr != nil; curr = curr.Next() {
+		task := curr.Value.(*Task)
+
+		if task.State != taskQueued {
+			// If task is already in process or finished, skip
+			continue
+		}
+		log.Debug(fmt.Sprintf("Textile sync [%s]: Processing task %s", queueName, task.Type))
+		task.State = taskPending
+
+		handleExecResult := func(err error) {
+			if err == nil {
+				// Task completed successfully
+				log.Debug(fmt.Sprintf("Textile sync [%s]: task completed succesfully", queueName))
+			} else {
+				log.Error(fmt.Sprintf("Textile sync [%s]: task failed", queueName), err)
+			}
+		}
+
+		if task.Parallelizable && parallelTaskCount < maxParallelTasks {
+			parallelTaskCount++
+
+			go func() {
+				err := s.executeTask(ctx, task)
+				handleExecResult(err)
+				parallelTaskCount--
```

Comment:
@dmerrill6 

Do you think this is a race condition? I think that [atomic](https://golang.org/pkg/sync/atomic/) would be helpful here.

Reply:
There shouldn't be any race condition issues as `sync` func is behind a lock. However after this comment I noticed this var was being shared by both queues so I fixed that in 10b65a3